### PR TITLE
Artifact and Packaging reorganization post 0.9.0

### PR DESCRIPTION
**Pull request by raulraja**
# Starting in Arrow 0.9.0 

Starting in 0.9.0 Arrow follows the following convention for artifact publication.

The arrow modules are Core, Effects, Optics, Recursion, etc.

An Arrow module is composed of data types and type classes.
Arrow modules are exported and published with the following semantics.

If we take for example `arrow-core`. 

Arrow core contains the basic arrow type classes and data types and it's composed of 3 main artifacts that may be used a la carte:

Recommended for most use cases:

- `arrow-core` (Depends on data and extensions modules and exports both)

Trimmed down versions:

- `arrow-core-data` (Only data types)
- `arrow-core-extensions` (Only type class extensions)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,35 +40,59 @@ allprojects {
 }
 ```
 
+# Starting in Arrow 0.9.0
```
Comment (pakoito) [inline README.md]: Dependency breakdown starting in Arrow 0.9.0

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/either.kt [hunk 2]
```diff
@@ -1,15 +1,14 @@
 @file:Suppress("UnusedImports")
```
Comment (pakoito) [inline modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/either.kt]: Who is this for? All of them should be used here.

File: modules/core/arrow-extras-extensions/src/main/kotlin/arrow/data/extensions/store.kt [hunk 3]
```diff
@@ -1,10 +1,10 @@
-package arrow.instances
+package arrow.data.extensions
 
 import arrow.Kind
 import arrow.data.Store
 import arrow.data.StorePartialOf
 import arrow.data.fix
-import arrow.deprecation.ExtensionsDSLDeprecated
+
```
Comment (pakoito) [inline modules/core/arrow-extras-extensions/src/main/kotlin/arrow/data/extensions/store.kt]: Formatting here?

Comment (pakoito) [reply modules/core/arrow-extras-extensions/src/main/kotlin/arrow/data/extensions/store.kt]: nvm

File: modules/docs/arrow-docs/docs/docs/README.md [hunk 4]
```diff
@@ -45,34 +45,58 @@ allprojects {
 }
 ```
 
+# Starting in Arrow 0.9.0
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/README.md]: Same as above.

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/either.kt (hunk 2 above)
Comment (raulraja) [reply modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/either.kt]: Apparently a detekt exception recently introduced in a different PR

File: modules/docs/arrow-docs/docs/docs/generic/product/README.md [hunk 5]
```diff
@@ -96,12 +96,12 @@ hListOf(1000, 900).toAccount()
...
 
 ```kotlin:ank
-import arrow.instances.*
+import arrow.core.extensions.option.applicative.applicative
 
 val maybeBalance: Option<Int> = Option(1000)
 val maybeAvailable: Option<Int> = Option(900)
 
-ForOption extensions { 
+Option.applicative().run {
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/generic/product/README.md]: You should be able to call `Option.applicative().mapToAccount(maybeBalance, maybeAvailable)` directly here.

File: modules/docs/arrow-docs/docs/docs/generic/product/README.md [hunk 6]
```diff
@@ -110,16 +110,18 @@ ForOption extensions {
 val maybeBalance: Option<Int> = Option(1000)
 val maybeAvailable: Option<Int> = None
 
-ForOption extensions { 
+Option.applicative().run {
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/generic/product/README.md]: Same as above

File: modules/docs/arrow-docs/docs/docs/generic/product/README.md [hunk 7]
```diff
@@ -128,19 +130,20 @@ ForTry extensions {
...
-import arrow.effects.*
+import arrow.effects.coroutines.*
+import arrow.effects.coroutines.extensions.deferredk.applicative.applicative
 import kotlinx.coroutines.async
 
 val asyncBalance: DeferredK<Int> = DeferredK { 1000 }
 val asyncAvailable: DeferredK<Int> = DeferredK { 900 }
 
-ForDeferredK extensions { 
+DeferredK.applicative().run {
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/generic/product/README.md]: Same

File: modules/docs/arrow-docs/docs/docs/generic/product/README.md (hunk 5 above)
Comment (raulraja) [reply modules/docs/arrow-docs/docs/docs/generic/product/README.md]: `generic` does not export extensions for all the applicative data types, just polymorphic functions that can be used in the scope of applicative.

File: modules/docs/arrow-docs/docs/docs/generic/product/README.md (hunk 6 above)
Comment (raulraja) [reply modules/docs/arrow-docs/docs/docs/generic/product/README.md]: https://github.com/arrow-kt/arrow/pull/1226#discussion_r244870364

File: modules/docs/arrow-docs/docs/docs/generic/product/README.md (hunk 7 above)
Comment (raulraja) [reply modules/docs/arrow-docs/docs/docs/generic/product/README.md]: https://github.com/arrow-kt/arrow/pull/1226#discussion_r244870364

File: modules/docs/arrow-docs/docs/docs/patterns/dependencyinjection/README.md [hunk 8]
```diff
@@ -81,15 +81,18 @@ fun <F> Functor<F>.multiplyBy2(fa: Kind<F, Int>): Kind<F, Int> =
 And we can call it on the typeclass instances:
 
 ```kotlin:ank
-import arrow.instances.*
+import arrow.core.extensions.*
+import arrow.core.extensions.option.functor.functor
 
-ForOption extensions { 
+Option.functor().run {
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/patterns/dependencyinjection/README.md]: `Option.functor().multiplyBy2(Some(1))`

File: modules/free/arrow-free-extensions/build.gradle [hunk 9]
```diff
@@ -0,0 +1,15 @@
+dependencies {
+    compile "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlinVersion"
+    compile project(':arrow-annotations')
+    compile project(':arrow-free-data')
```
Comment (pakoito) [inline modules/free/arrow-free-extensions/build.gradle]: Maybe we can slim this down to just the extensions?

File: modules/recursion-schemes/arrow-recursion/README.md [hunk 10]
```diff
@@ -1,5 +0,0 @@
-# Recursion schemes for Arrow
```
Comment (pakoito) [inline modules/recursion-schemes/arrow-recursion/README.md]: This file should be gone altogether, I don't know how it's survived :D

File: settings.gradle [hunk 11]
```diff
@@ -29,50 +29,60 @@ buildscript {
...
-            _ 'data'
-            _ 'free'
-            _ 'instances-free'
-            _ 'instances-core'
-            _ 'instances-data'
+            _ 'meta'
+        }
+
+        core {
+            _ 'core-data'
```
Comment (pakoito) [inline settings.gradle]: core-core isn't here, whereas the others are, i.e. free-free or effects-effects

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/either.kt (hunk 2 above)
Comment (pakoito) [reply modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/either.kt]: Not for this PR then.

**Pull request merged**